Status: 2023Q3: SquashFS report (#274)
freebsd/freebsd-doc: website/content/en/status/report-2023-07-2023-09/squashfs.adoc

=== SquashFS driver for FreeBSD kernel

> Links: +
> link:https://wiki.freebsd.org/SummerOfCode2023Projects/PortSquashFuseToTheFreeBSDKernel[Wiki page] URL: 
> link:https://wiki.freebsd.org/SummerOfCode2023Projects/PortSquashFuseToTheFreeBSDKernel[] +
> link:https://github.com/Mashijams/freebsd-src/tree/gsoc/testing[Source code] URL: https://github.com/Mashijams/freebsd-src/tree/gsoc/testing[]

Reviewer: ```suggestion
Links: +
link:https://wiki.freebsd.org/SummerOfCode2023Projects/PortSquashFuseToTheFreeBSDKernel[Wiki page] URL: link:https://wiki.freebsd.org/SummerOfCode2023Projects/PortSquashFuseToTheFreeBSDKernel[] +
link:https://github.com/Mashijams/freebsd-src/tree/gsoc/testing[Source code] URL: link:https://github.com/Mashijams/freebsd-src/tree/gsoc/testing[]
```

Contact: Raghav Sharma <raghav@FreeBSD.org>

> This quarter we finished SquashFS driver work for kernel. We now can mount SquashFS archives on the 13.2 or greater FreeBSD release and perform all basic 
> read-only filesystem operations.

Reviewer: ```suggestion
This quarter we finished SquashFS driver work for the kernel.
We now can mount SquashFS archives on FreeBSD 13.2-RELEASE or greater, and perform all basic read-only filesystem operations.
```

Code work includes:

* Implementing vop_lookup() and vop_readdir() hooks for directory read support.
* Implementing vop_read() and vop_strategy() hooks for files read support.
* Implementing vop_readlink() hook for symlinks read support.

> We also implemented extended attributes interface functions for SquashFS and all the work remaining for them is to implement their kernel interface hooks.

Reviewer: ```suggestion
We also implemented extended attributes interface functions for SquashFS.
All that remains is to implement their kernel interface hooks.
```

> There were a lot of bug fixes as well, one major one was to find out why we can't list the first entry of the root directory, turns out SquashFS could have inode_number as zero 
> which kernel for some reason skips while listing dirents.

Reviewer: ```suggestion
There were a lot of bug fixes as well.
One major issue was to find out why we can not list the first entry of the root directory, it transpires that SquashFS could have `inode_number` as zero, which the kernel, for some reason, skips while listing dirents.
```

> For now, we fixed it by passing dummy inode_number to dirent instead of zero.

Reviewer: ```suggestion
For now, we fixed it by passing dummy `inode_number`, instead of zero, to dirent.
```

The code review is currently ongoing with my mentor mailto:chuck@FreeBSD.org[Chuck Tuffli].

> I'm happy to say that SquashFS will find its place in upcoming FreeBSD releases.

Reviewer: ```suggestion
I am happy to say that SquashFS will find its place in upcoming FreeBSD releases.
```

Reviewer: This formality is according to [a guideline for the FreeBSD Documentation Project](https://docs.freebsd.org/en/books/fdp-primer/book/#writing-style-guidelines).
